### PR TITLE
#162949075 Users should be able to create their profiles

### PR DESCRIPTION
**Pull request by sharkdevs**
### Getting Started

1. Open terminal and clone the repo:

    `git clone https://github.com/andela/ah-robotics.git`

2. Change directory into the cloned repository.

    `cd ah-robotics`

3. Checkout to the Feature branch

    `git checkout ft-user-profile-162949075`

4. Create a virtual environment by running the following commands:

    `virtualenv -p python3 env`

    or

    `python3 -m virtualenv env`

5. Once the virtual environment has been setup, activate it and install the packages required by the project,
    these are contained in the file named `requirements.txt`

      `source env/bin/activate`

      `pip install -r requirements.txt`

### setting up the database

> Ensure PostgreSQL is installed and is running psql, follow these steps to create a new user and a database which will be used in the application.

  1. Start the postgres database server using the command:
  
      `pg_ctl -D /usr/local/var/postgres start`

  2. Run psql interactive terminal

      `psql postgres`

  3. Create a new user:

      `CREATE USER sample_username WITH PASSWORD 'sample_password';`

  4. Grant privileges to the user:

      `ALTER USER sample_username CREATEDB;`

  5. Create a database:

      `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
       
  6.   Exit the psql shell using `\q`

  7. Create a .env as guided by the .env_example and add the following configuration variables

      ```source-json
      source env/bin/activate
      export DB_NAME="database_name"
      export DB_USER="sample_username"
      export DB_PASS="sample_password"
      export DB_HOST="host_name_eg_localhost"
      export DB_PORT="port_number_eg_5432"
      export TIME_DELTA="30"
      export EMAIL_HOST="smtp.gmail.com"
      export EMAIL_HOST_USER="user@example.com"
      export EMAIL_HOST_PASSWORD="email_password"
      export EMAIL_PORT=587
      export PASSWORD_RESET_URL="api/v1/reset_password"
      export CLIENT_DOMAIN="127.0.0.1:8000"
      export CLIENT_DOMAIN="127.0.0.1:8000"
      export CLOUDINARY_NAME="your_cloudinary_name"
      export CLOUDINARY_KEY="your_cloudinary_key"
      export CLOUDINARY_SECRET="your_secret_key"
      ```
  8. Source .env to set the configuration variables:

      `source .env`
  9. Run migrations to create the tables in the database:

      ```source-json
      python manage.py makemigrations
      python manage.py migrate
      ```

---
### Collect static files

Run the following command to collect static files.

`python manage.py collectstatic`

### Run The Server

Once everything has been setup, to run the application run the command:

   `python manage.py runserver`

The application can be accessed in the url <http://localhost:8000>

---

### Testing the feature

1. Run the server from the terminal using python manage.py runserver.

2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL <http://127.0.0.1:8000/api/v1/users/> to register a user.
3. On the body section on postman select `raw` and type as `JSON/application`.
Add your payload to the body section using the following format `but NOT with the exact credentials`:

> Kindly use a genuine email you have access to because an email will be sent, do NOT use a real password.

   ```
    http://127.0.0.1:8000/api/v1/users/
 ```
        {
        "user": {
            "username": "dojo1234",
            "email": "user@example.com",
            "password": "test1234#"
            }
          }
     
4. Once all the fields are inserted, click POST to register the user successfully.
5. Repeat this step again to register another user.
6. Change the endpoint to <http://127.0.0.1:8000/api/v1/users/login/>
7. On the body section on postman select `raw` and type as `JSON/application`. Add the following details with the credentials you registered a user with:

    ```
     http://127.0.0.1:8000/api/v1/users/login/
   ```

    ```
        {
        "user": {
            "email": "user@example.com",
            "password": "test1234#"
            }
          }
     ```
8. Send the Request and copy the token section of the response gotten.
9. Change to a `GET` request and add the endpoint: <http://127.0.0.1:8000/api/v1/profiles/> to view search bar.
10. Click the tab labeled `Headers` and add information as in the screenshot below. Ensure you type the word token before pasting you token in the authorization row.
> Click the first row and add `content-type`. Match it in adjacent column same row with `application/json`.
> in the second row, type `Authorizations` and in the corresponding field, type `token` then paste your token.
screenshot 2019-01-24 at 22 18 08
11. To view a specific user use this url format below: 

   
   ``` 
http://127.0.0.1:8000/api/v1/profiles/<sample_user_name/>
 ```
    
12. To update the user profile, change the request to `PATCH`  
   ```
<http://127.0.0.1:8000/api/v1/profiles/sample_user_name/>
```
In the body section use the form data to update fields(bio and image)

screenshot 2019-01-25 at 12 56 40

### Running tests
1. Go to the terminal and ensure that the server is not running.
2. Run the tests 
 `python manage.py test`

### What are the relevant pivotal tracker stories?
[#162949075](https://www.pivotaltracker.com/n/projects/2235171/stories/162949075)

Screenshots:
1. User Registration triggers profile creation.

screen shot 2019-01-24 at 19 22 49

2. Login the user
screen shot 2019-01-24 at 19 24 05

3. View all profiles
screen shot 2019-01-24 at 19 25 49

4. View one profile
screen shot 2019-01-24 at 19 26 38

5. Update my profile
screenshot 2019-01-25 at 12 56 40

### CODE REVIEW

**Comment by Teatoller** [comment]
Hi Sharkdevs/Joma

I am at step "6" in setting up the database.
Where or how do I obtain cloudinary params?

```
export CLOUDINARY_NAME="your_cloudinary_name"
export CLOUDINARY_KEY="your_cloudinary_key"
export CLOUDINARY_SECRET="your_secret_key"
```
Thank you joma,
May you flesh the points 9-11 "Testing the Feature. You have left out small steps that can be helpful for a beginner to follow through.
Regards
Steven Ennis

**Comment by jomasim** [comment]
> Hi Sharkdevs/paulstar200
> 
> I am at step "6" in setting up the database.
> Where or how do I obtain cloudinary params?
> 
> ```
> export CLOUDINARY_NAME="your_cloudinary_name"
> export CLOUDINARY_KEY="your_cloudinary_key"
> export CLOUDINARY_SECRET="your_secret_key"
> ```
> Regards
> Steven Ennis

@Teatoller  I shared cloudinary group account login credentials on slack for the just check.

**Review by Paulvitalis200 — COMMENTED** [review]
Hey @sharkdevs and @jomasim ,
I was unable to proceed from step 11 in the pull Request. Please review your Pull Request instructions.

**Review by waracci — COMMENTED** [review]
Kindly check on how to include links in the PR message otherwise it looks good to me.
Here is a great resource [Markdown guidelines](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links)

**Review by Paulvitalis200 — APPROVED** [review]
Hey @sharkdevs and @waracci ,
Thank you for implementing the changes.  I have been able to test it on my machine and it works as expected. Good work.

**Review by johnwayodi — APPROVED** [review]
Nice work. I managed to upload an image and update the user profile. The tests have passed on my machine.

File: authors/apps/profiles/tests/base_test.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                "email": "2016@gmail.com",
+                "password": "manu232#$$"
+
+            }
+        }
+        self.client = APIClient()
+        self.login_url = reverse('authentication:login_url')
+        self.register_url = reverse('authentication:register_url')
+        self.profiles_url = reverse('profiles:profiles')
+
```
Comment (Collins33) [inline authors/apps/profiles/tests/base_test.py]: Why do we have a lot of white space here?

File: authors/apps/profiles/tests/base_test.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            url, 
+            self.user_bio, content_type="application/json")
+    
+    def authorize_user(self,user_details):
+        # register a user
+        self.register_user(data=self.user)
+        payload = self.login_a_user(data=user_details)
+        self.client.credentials(HTTP_AUTHORIZATION= 'token '+ payload['token'])
+
+    # def forgot_password_req(self, data):
```
Comment (Collins33) [inline authors/apps/profiles/tests/base_test.py]: Remove commented out code

File: authors/apps/profiles/tests/test_profiles.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        """Test that a user can view all user profiles"""
+        self.authorize_user(self.user_login_details)
+        response = self.client.get(self.profiles_url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_that_a_user_can_edit_their_profile(self):
+        """Tests the edit profile functionality"""
+        self.authorize_user(self.user_login_details)
+
+    def test_that_a_user_cannot_edit_another_users_profile(self):
```
Comment (Collins33) [inline authors/apps/profiles/tests/test_profiles.py:37]: Is this testing anything?

File: authors/apps/profiles/views.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            user_profile = UserProfile.objects.get(user__username=username)
+        except:
+            response = {"error": "User profile does not exist."}
+            return Response(response, status=status.HTTP_404_NOT_FOUND)
+        serializer = self.serializer_class(user_profile)
+        return Response({
+            'profile': serializer.data}, status=status.HTTP_200_OK)
+
+    def patch(self, request, username):
+        # This method updates users to update their own profiles
```
Comment (Collins33) [inline authors/apps/profiles/views.py]: Use a multiline comment for this

File: authors/apps/core/models.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+class TimestampMixin(models.Model):
+    """
+    This model creates created_at and updated_at fields
+    """
+    created_at = models.DateField(auto_now_add=True)
+    updated_at = models.DateField(auto_now=True)
+
+    class Meta:
+        # make the model to prevent defualt migration but instead adds
```
Comment (Collins33) [inline authors/apps/core/models.py]: Fix the typo for the word default

File: authors/apps/profiles/models.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    @property
+    def username(self):
+        return self.user.username
+
+    def __str__(self):
+        return self.user.username
+
+    @receiver(post_save, sender=User)
+    def create_user_profile(sender, instance, created, **kwargs):
+        if created:
```
Comment (Collins33) [inline authors/apps/profiles/models.py:35]: Write a description as to what create_user_profile function does and what save_user_profile does

File: authors/apps/profiles/tests/test_profiles.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    def test_view_all_users_profiles(self):
+        """Test that a user can view all user profiles"""
+        self.authorize_user(self.user_login_details)
+        response = self.client.get(self.profiles_url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_that_a_user_can_edit_their_profile(self):
+        """Tests the edit profile functionality"""
+        self.authorize_user(self.user_login_details)
```
Comment (Collins33) [inline authors/apps/profiles/tests/test_profiles.py:30]: Is this testing anything?

**Review by Collins33 — CHANGES_REQUESTED** [review]

File: authors/apps/profiles/models.py (hunk 6 above)
Comment (jomasim) [reply authors/apps/profiles/models.py:35]: @Collins33  I have added the description to create_user_profile and save_user_profile methods

File: authors/apps/profiles/tests/test_profiles.py (hunk 7 above)
Comment (jomasim) [reply authors/apps/profiles/tests/test_profiles.py:30]: The test was incomplete I have managed to add  two assertions
- to test whether the payload contains the updated bio info
- to test the status code when updating a user profile

File: authors/apps/profiles/tests/test_profiles.py (hunk 3 above)
Comment (jomasim) [reply authors/apps/profiles/tests/test_profiles.py:37]: This test also was not complete I have added necessary assertions.

File: authors/apps/core/models.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+class TimestampMixin(models.Model):
+    """
+    This model creates created_at and updated_at fields
+    """
+    created_at = models.DateField(auto_now_add=True)
+    updated_at = models.DateField(auto_now=True)
+
+    class Meta:
+        """
+        make the model to prevent default migration but instead adds
```
Comment (jomasim) [inline authors/apps/core/models.py:13]: @Collins33  fixed the typo in the word  'default'

File: authors/apps/profiles/tests/base_test.py (hunk 2 above)
Comment (jomasim) [reply authors/apps/profiles/tests/base_test.py]: I  have removed the commented code.

File: authors/apps/profiles/views.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+            user_profile = UserProfile.objects.get(user__username=username)
+        except:
+            response = {"error": "User profile does not exist."}
+            return Response(response, status=status.HTTP_404_NOT_FOUND)
+        serializer = self.serializer_class(user_profile)
+        return Response({
+            'profile': serializer.data}, status=status.HTTP_200_OK)
+
+    def patch(self, request, username):
+        """
```
Comment (jomasim) [inline authors/apps/profiles/views.py]: @Collins33  I have included multiline comment on the patch method.

**Comment by Teatoller** [comment]
Hi Joma/Sharkdevs
This is a good piece of work.
Kudos

Regards
Steven Ennis

**Review by Collins33 — APPROVED** [review]

**Comment by Collins33** [comment]
The pull request works as expected. Great job!!!

**Pull request merged**